### PR TITLE
[CSM] Change command prefix to "." and add "help" command.

### PR DESCRIPTION
**Pull request by red-001**
Setting the prefix to "/" makes writing client-sided mods harder because you risk blocking access to server-sided commands. It also makes it harder to add a "help" command as a different name has to be used.

### CODE REVIEW

**Comment by ShadowNinja** [comment]
It looks like `.help` might be copy/pasted from the server-side version.  I'd remove that duplication if possible.
Otherwise, LGTM.

**Comment by nerzhul** [comment]
@ShadowNinja i don't know if it's possible to factorize core for this container, can you try to look at this @red-001 ?

**Comment by red-001** [comment]
@ShadowNinja done.

File: builtin/common/chatcommands.lua [hunk 1]
```diff
@@ -27,4 +27,70 @@ function core.override_chatcommand(name, redefinition)
 		rawset(chatcommand, k, v)
 	end
 	core.registered_chatcommands[name] = chatcommand
-end
\ No newline at end of file
+end
+
+local function priv_cmd(name, param)
+	local function format_help_line(cmd, def)
+		local msg = core.colorize("#00ffff", "/"..cmd)
```
Comment (nerzhul) [inline builtin/common/chatcommands.lua]: / or . no ?

Comment (red-001) [inline builtin/common/chatcommands.lua]: fixed

File: builtin/common/chatcommands.lua [hunk 2]
```diff
@@ -27,4 +27,76 @@ function core.override_chatcommand(name, redefinition)
 		rawset(chatcommand, k, v)
 	end
 	core.registered_chatcommands[name] = chatcommand
-end
\ No newline at end of file
+end
+
+local prefix = "/"
```
Comment (nerzhul) [inline builtin/common/chatcommands.lua]: cmd_marker is better than just "prefix"

Comment (red-001) [inline builtin/common/chatcommands.lua]: done

File: builtin/common/chatcommands.lua [hunk 3]
```diff
@@ -27,4 +27,76 @@ function core.override_chatcommand(name, redefinition)
...
+		else
+			return true, format_help_line(cmd, def)
+		end
+	end
+end
+
+if INIT == "client" then
+	core.register_chatcommand("help", {
+		params = "[all/<cmd>]",
+		description = "Get help for commands or list privileges",
```
Comment (SmallJoker) [inline builtin/common/chatcommands.lua]: Does not list privileges.

File: builtin/client/chatcommands.lua [hunk 4]
```diff
@@ -2,27 +2,35 @@
...
 core.register_on_sending_chat_messages(function(message)
-	if not (message:sub(1,1) == "/") then
-		return false
+	local first_char = message:sub(1,1)
+	if first_char == "/" or first_char == "." then
+		core.display_chat_message("issued command: " .. message)
 	end
 
-	core.display_chat_message("issued command: " .. message)
+	if first_char ~= "." then
```
Comment (SmallJoker) [inline builtin/client/chatcommands.lua]: How about `core.setting_get("keymap_cmd_local")`?

Comment (red-001) [inline builtin/client/chatcommands.lua]: iirc `setting_get` is currently unavailable on the client.

Comment (SmallJoker) [inline builtin/client/chatcommands.lua]: Alright, should be replaced as soon these functions are available.

File: src/defaultsettings.cpp [hunk 5]
```diff
@@ -71,6 +71,7 @@ void set_default_settings(Settings *settings)
 	settings->setDefault("keymap_special1", "KEY_KEY_E");
 	settings->setDefault("keymap_chat", "KEY_KEY_T");
 	settings->setDefault("keymap_cmd", "/");
+	settings->setDefault("keymap_cmd_local", ".");
```
Comment (SmallJoker) [inline src/defaultsettings.cpp]: Needs a minetest.conf entry.

Comment (nerzhul) [inline src/defaultsettings.cpp]: i don't think it's a good idea as it should be a standard command for all users

Comment (red-001) [inline src/defaultsettings.cpp]: @SmallJoker  Done.
@nerzhul other key mappings are in `minetest.conf.example` and `settingtypes.txt`.

File: builtin/common/chatcommands.lua [hunk 6]
```diff
@@ -27,4 +27,76 @@ function core.override_chatcommand(name, redefinition)
...
+			return true, format_help_line(cmd, def)
+		end
+	end
+end
+
+if INIT == "client" then
+	core.register_chatcommand("help", {
+		params = "[all/<cmd>]",
+		description = "Get help for commands",
+		func = function(param)
```
Comment (rubenwardy) [inline builtin/common/chatcommands.lua]:
```lua 
local func = priv_cmd
if INIT == "client" then
	func = function(param)
		return priv_cmd(nil, param)
	end
end

core.register_chatcommand("help", {
	params = "[all/privs/<cmd>]",
	description = "Get help for commands or list privileges",
	func = priv_cmd,
})
```

File: builtin/common/chatcommands.lua [hunk 7]
```diff
@@ -27,4 +27,76 @@ function core.override_chatcommand(name, redefinition)
...
\ No newline at end of file
+end
+
+local cmd_marker = "/"
+
+if INIT == "client" then
+	cmd_marker = "."
+end
+
+local function priv_cmd(name, param)
```
Comment (rubenwardy) [inline builtin/common/chatcommands.lua]: `do_help_cmd`?

File: builtin/common/chatcommands.lua [hunk 8]
```diff
@@ -27,4 +27,76 @@ function core.override_chatcommand(name, redefinition)
...
+	elseif param == "all" then
+		local cmds = {}
+		for cmd, def in pairs(core.registered_chatcommands) do
+			if INIT == "client" or core.check_player_privs(name, def.privs) then
+				cmds[#cmds + 1] = format_help_line(cmd, def)
+			end
+		end
+		table.sort(cmds)
+		return true, "Available commands:\n"..table.concat(cmds, "\n")
+	elseif INIT == "game" and param == "privs" then
```
Comment (rubenwardy) [inline builtin/common/chatcommands.lua]: whilst it's potentially out of scope, the code above here can be simplified:

```lua
if param == "" or param == "all" then
	local cmds = {}
	for cmd, def in pairs(core.registered_chatcommands) do
		if INIT == "client" or core.check_player_privs(name, def.privs) then
			cmds[#cmds + 1] = cmd
		end
	end
	table.sort(cmds)

	if param == "" then
		return true, "Available commands: " .. table.concat(cmds, " ") .. "\n"
				.. "Use '"..cmd_marker.."help <cmd>' to get more information,"
				.. " or '"..cmd_marker.."help all' to list everything."
	else
		return true, "Available commands:\n"..table.concat(cmds, "\n")
	end

```

File: builtin/common/chatcommands.lua (hunk 6 above)
Comment (red-001) [inline builtin/common/chatcommands.lua]: it wouldn't make much sense to do that as the params and the description is different

Comment (rubenwardy) [inline builtin/common/chatcommands.lua]: oh, didn't notice that

**Comment by sofar** [comment]
`.` is fairly decent. I've seen it used as an alternate cmd prefix character in several programs before.

File: builtin/common/chatcommands.lua (hunk 7 above)
Comment (red-001) [inline builtin/common/chatcommands.lua]: Done.

**Comment by red-001** [comment]
Squashed, renamed `privs_cmd` to `do_help_cmd` and removed an unused variable.

File: builtin/common/chatcommands.lua (hunk 8 above)
Comment (red-001) [inline builtin/common/chatcommands.lua]: "all" stores the result of `format_help_line` in `cmds` not the name of the command.

**Comment by paramat** [comment]
e70e15134c95d37241bb6f6124105c0f1c08ab8a

**Pull request closed without merging**